File: android/src/main/java/com/hlabexmaples/easyapi/ui/UsersFragment.java
package com.hlabexmaples.easyapi.ui;

import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.view.View;
import com.hlab.easyapi_java.easyapi.main.api.EasyApi;
import com.hlab.easyapi_java.easyapi.main.api.EasyApiCall;
import com.hlab.easyapi_java.easyapi.util.NetworkUtils;
import com.hlabexmaples.easyapi.R;
import com.hlabexmaples.easyapi.data.models.Envelop;
import com.hlabexmaples.easyapi.data.models.User;
import com.hlabexmaples.easyapi.data.webservice.ApiFactory;
import com.hlabexmaples.easyapi.databinding.FragmentUsersBinding;
import com.hlabexmaples.easyapi.ui.base.BaseFragment;
import io.reactivex.observers.DisposableObserver;
import java.util.List;
import java.util.Objects;

/**
 * Created by H.T. on 22/02/18.
 */

public class UsersFragment extends BaseFragment<FragmentUsersBinding>
    implements TabLayout.OnTabSelectedListener {

  private UserListAdapter adapter;
  private EasyApiCall<Envelop<List<User>>> usersListApi;
  private EasyApiCall<Envelop<List<User>>> usersListRxApi;

  @Override
  protected int defineLayoutResource() {
    return R.layout.fragment_users;
  }

  @Override
  protected void initializeComponent(View view) {
    //EasyApi Configuration
    usersListApi = new EasyApi.Builder<Envelop<List<User>>>(Objects.requireNonNull(getActivity()))
        .setLoaderInterface(this).build();
    usersListRxApi = new EasyApi.Builder<Envelop<List<User>>>(Objects.requireNonNull(getActivity()))
        .setLoaderInterface(this).configureWithRx().build();

    adapter = new UserListAdapter();
    binding.rvUsers.setAdapter(adapter);

    binding.tab.addOnTabSelectedListener(this);

    fetchUsers();
  }

  @Override
  protected void initToolbar() {
    assert getActivity() != null;
    binding.toolbar.setTitle(getString(R.string.title_users));
    binding.toolbar.setNavigationIcon(R.drawable.ic_back);
    binding.toolbar.setNavigationOnClickListener(view -> goBack());
  }

  @Override
  public void onClick(View v) {

  }

  /**
   * Example 1:
   * Load via normal retrofit call
   */
  private void fetchUsers() {
    usersListApi
        .initCall(ApiFactory.getInstance().fetchUsers("1"))
        .execute(true, (response, isError) -> {
          if (!isError) {
            adapter.setUserList(response.getData());
          }
        });
  }

  /**
   * Example 2: (default)
   * Load via normal RxJava call
   */
  private void fetchUsersRx() {
    usersListRxApi.initCall(ApiFactory.getInstance().fetchUsersWithRx("1"))
        .execute(true, (response, isError) -> {
          if (!isError) {
            adapter.setUserList(response.getData());
          }
        });
  }

  /**
   * Example 3:
   * Load via custom RxJava observer
   */
  private void fetchUsersRxExample1() {
    DisposableObserver<Envelop<List<User>>> observer = getExample1Observer();

    showLoading();
    usersListRxApi.initCall(ApiFactory.getInstance().fetchUsersWithRx("1"))
        .executeWith(observer);
  }

  /**
   * Example 4:
   * Load via RxJava Consumer Callbacks
   */
  private void fetchUsersRxExample2() {
    showLoading();
    usersListRxApi.initCall(ApiFactory.getInstance().fetchUsersWithRx("1"))
        .executeWith(value -> {
          hideLoading();
          if (value.getData() != null) {
            adapter.setUserList(value.getData());
          } else {
            showError(getString(R.string.alert_message_error));
          }
        }, e -> {
          hideLoading();
          NetworkUtils.handleErrorResponse(e);
        });
  }

  @NonNull
  private DisposableObserver<Envelop<List<User>>> getExample1Observer() {
    return new DisposableObserver<Envelop<List<User>>>() {

      @Override
      public void onNext(Envelop<List<User>> value) {
        if (value.getData() != null) {
          adapter.setUserList(value.getData());
        } else {
          showError(getString(R.string.alert_message_error));
        }
      }

      @Override
      public void onError(Throwable e) {
        NetworkUtils.handleErrorResponse(e);
      }

      @Override
      public void onComplete() {
        hideLoading();
      }
    };
  }

  @Override
  public void showLoading() {
    binding.setIsLoading(true);
  }

  @Override
  public void hideLoading() {
    binding.setIsLoading(false);
  }

  @Override
  public void onTabSelected(TabLayout.Tab tab) {
    int tag = tab.getPosition();
    switch (tag) {
      case 0:
        fetchUsers();
        break;
      case 1:
        fetchUsersRx();
        break;
    }
  }

  @Override
  public void onTabUnselected(TabLayout.Tab tab) {

  }

  @Override
  public void onTabReselected(TabLayout.Tab tab) {

  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    usersListApi.dispose();
    usersListRxApi.dispose();
  }
}

File: kotlin/src/main/java/com/hlabexamples/easyapi/kotlin/ui/UsersFragment.kt
package com.hlabexamples.easyapi.kotlin.ui

import android.support.design.widget.TabLayout
import android.view.View
import com.hlab.easyapi_kotlin.easyapi.main.api.EasyApiBase.Builder
import com.hlab.easyapi_kotlin.easyapi.main.api.EasyApiCall
import com.hlab.easyapi_kotlin.easyapi.util.NetworkUtils
import com.hlabexamples.easyapi.kotlin.data.models.Envelop
import com.hlabexamples.easyapi.kotlin.data.models.User
import com.hlabexamples.easyapi.kotlin.data.webservice.ApiFactory
import com.hlabexamples.easyapi.kotlin.ui.base.BaseFragment
import com.hlabexmaples.easyapi.R
import com.hlabexmaples.easyapi.databinding.FragmentUsersBinding
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import io.reactivex.observers.DisposableObserver

/**
 * Created by H.T. on 22/02/18.
 */

class UsersFragment : BaseFragment<FragmentUsersBinding>(), TabLayout.OnTabSelectedListener {

  private var disposable: CompositeDisposable? = null
  private var adapter: UserListAdapter? = null
  private var usersListApi: EasyApiCall<Envelop<List<User>>>? = null
  private var usersListRxApi: EasyApiCall<Envelop<List<User>>>? = null

  override fun defineLayoutResource(): Int {
    return R.layout.fragment_users
  }

  override fun initializeComponent(view: View) {
    disposable = CompositeDisposable()
    usersListApi = Builder<Envelop<List<User>>>(activity!!)
        .setLoaderInterface(this)
        .build()
    usersListRxApi =
        Builder<Envelop<List<User>>>(activity!!)
            .setLoaderInterface(this)
            .configureWithRx()
            .build()

    adapter = UserListAdapter()
    binding.rvUsers.adapter = adapter

    binding.tab.addOnTabSelectedListener(this)

    fetchUsers()

  }

  override fun initToolbar() {
    assert(activity != null)
    binding.toolbar.title = getString(R.string.title_users)
    binding.toolbar.setNavigationIcon(R.drawable.ic_back)
    binding.toolbar.setNavigationOnClickListener {
      goBack()
    }
  }

  override fun onClick(v: View) {

  }

  /**
   * Example 1:
   * Load via normal retrofit call
   */
  private fun fetchUsers() {
    usersListApi?.initCall(ApiFactory.instance!!.fetchUsers("1"))!!
        .execute(true) { response, isError: Boolean ->
          if (!isError)
            response?.data?.let { adapter?.setUserList(it) }
        }
  }

  /**
   * Example 2: (default)
   * Load via normal RxJava call
   */
  private fun fetchUsersRx() {
    usersListRxApi?.initCall(ApiFactory.instance!!.fetchUsersWithRx("1"))!!
        .execute(true) { response, isError ->
          if (!isError)
            response?.data?.let { adapter?.setUserList(it) }
        }
  }

  /**
   * Example 3:
   * Load via custom RxJava observer
   */
  private fun fetchUsersRxExample1() {
    val observer = getExample1Observer()

    showLoading()
    usersListRxApi?.initCall(ApiFactory.instance!!.fetchUsersWithRx("1"))!!
        .executeWith(observer)
  }

  /**
   * Example 4:
   * Load via RxJava Consumer Callbacks
   */
  private fun fetchUsersRxExample2() {
    showLoading()
    usersListRxApi?.initCall(ApiFactory.instance!!.fetchUsersWithRx("1"))!!
        .executeWith(Consumer { value ->
          hideLoading()
          if (value.data != null) {
            adapter?.setUserList(value.data!!)
          } else {
            showError(getString(R.string.alert_message_error))
          }
        }, Consumer { t ->
          hideLoading()
          showError(NetworkUtils.handleErrorResponse(t!!))
        })
  }

  private fun getExample1Observer(): DisposableObserver<Envelop<List<User>>> {
    return object : DisposableObserver<Envelop<List<User>>>() {

      override fun onNext(value: Envelop<List<User>>) {
        if (value.data != null) {
          adapter?.setUserList(value.data!!)
        } else
          showError(context!!.getString(R.string.alert_message_error))

      }

      override fun onError(e: Throwable) {
        NetworkUtils.handleErrorResponse(e)
      }

      override fun onComplete() {
        hideLoading()
      }
    }
  }

  override fun showLoading() {
    binding.isLoading = true
  }

  override fun hideLoading() {
    binding.isLoading = false
  }

  override fun onTabSelected(tab: TabLayout.Tab) {
    val tag = tab.position
    when (tag) {
      0 -> fetchUsers()
      1 -> fetchUsersRx()
    }
  }

  override fun onTabUnselected(tab: TabLayout.Tab) {

  }

  override fun onTabReselected(tab: TabLayout.Tab) {

  }

  override fun onDestroy() {
    super.onDestroy()
    if (disposable != null && !disposable!!.isDisposed) {
      disposable!!.dispose()
    }
  }
}
